### ISSUE TITLE
[Testing] Add Comprehensive Test Suite for Core Functionality

### ISSUE THREAD

**Issue by MolinMa** [H-EmbodVis/GRANT#2]
The GRANT project currently has minimal test coverage, with only low-level CUDA operation tests and no tests for core functionality including trainers, datasets, models, or LLM components. This limits the ability to safely refactor code, catch regressions, and ensure reproducibility.

### Existing Tests
**Location**: `utils/pointops2/functions/test_*.py`
**Coverage**: 8 test files for custom CUDA operations only
**Quality**: Good - uses proper seeding (`torch.manual_seed(1)`)

#### Critical Components Without Tests
1. ❌ **Trainer Logic** (trainer/trainer.py - 2,063 lines)
   - No tests for training loop
   - No tests for validation logic
   - No tests for metric computation

2. ❌ **Dataset Loading** (datasets/)
   - No tests for SceneVerse dataset
   - No tests for language annotation processing
   - No tests for data augmentation

3. ❌ **Model Components** (models/)
   - No tests for LLM integration
   - No tests for 3D encoder
   - No tests for prediction heads

4. ❌ **Utilities** (utils/)
   - No tests for data preprocessing
   - No tests for evaluation metrics
   - No tests for helper functions

5. ❌ **Integration Tests**
   - No end-to-end pipeline tests
   - No tests for checkpoint saving/loading
   - No tests for distributed training

### Development Impact
- **Refactoring Risk**: Cannot safely modify code without manual verification
- **Regression Detection**: No automated way to catch bugs introduced by changes
- **Debugging Time**: More time spent debugging issues that tests would catch
- **Code Confidence**: Uncertain if changes break existing functionality

### Research Impact
- **Reproducibility**: Difficult to verify consistent behavior across runs
- **Experimentation**: Risky to try new approaches without safety net
- **Collaboration**: Contributors hesitant to modify untested code
- **Paper Claims**: Hard to validate claims without systematic testing

### Production Impact
- **Deployment Risk**: No confidence in code correctness
- **Bug Discovery**: Issues only found in production
- **Maintenance Cost**: Higher cost to fix bugs without tests

### Testing Pyramid

```
       /\
      /  \  E2E Tests (5%)
     /____\
    /      \  Integration Tests (25%)
   /________\
  /          \  Unit Tests (70%)
 /__________\
```

#### 1.1 Dataset Tests

```python
# tests/test_datasets/test_sceneverse.py
import pytest
import torch
from datasets.sceneverse import SceneVerseDataset

class TestSceneVerseDataset:
    """Test SceneVerse dataset loading and processing."""

    @pytest.fixture
    def dataset(self):
        """Create a test dataset instance."""
        config = {
            'data_root': 'tests/fixtures/data',
            'split': 'test',
            'max_obj_len': 200
        }
        return SceneVerseDataset(config)

    def test_dataset_length(self, dataset):
        """Test dataset returns correct number of samples."""
        assert len(dataset) > 0
        assert isinstance(len(dataset), int)

    def test_getitem_returns_valid_format(self, dataset):
        """Test __getitem__ returns expected data format."""
        sample = dataset[0]

        # Check all required keys exist
        assert 'point_clouds' in sample
        assert 'lang_tokens' in sample
        assert 'target_ids' in sample

        # Check data types
        assert isinstance(sample['point_clouds'], torch.Tensor)
        assert isinstance(sample['lang_tokens'], torch.Tensor)

    def test_point_cloud_shape(self, dataset):
        """Test point cloud has correct dimensions."""
        sample = dataset[0]
        pc = sample['point_clouds']

        # Should be (N, 3+features)
        assert pc.ndim == 2
        assert pc.shape[1] >= 3  # At least XYZ

    def test_language_token_range(self, dataset):
        """Test language tokens are within valid vocabulary range."""
        sample = dataset[0]
        tokens = sample['lang_tokens']

        # Tokens should be non-negative integers
        assert torch.all(tokens >= 0)
        assert tokens.dtype == torch.long

    def test_deterministic_with_seed(self):
        """Test dataset returns same data with fixed seed."""
        torch.manual_seed(42)
        dataset1 = SceneVerseDataset({'data_root': 'tests/fixtures/data'})
        sample1 = dataset1[0]

        torch.manual_seed(42)
        dataset2 = SceneVerseDataset({'data_root': 'tests/fixtures/data'})
        sample2 = dataset2[0]

        assert torch.allclose(sample1['point_clouds'], sample2['point_clouds'])

    def test_handles_missing_data_gracefully(self, dataset):
        """Test dataset handles missing or corrupted data."""
        # Test with invalid index
        with pytest.raises(IndexError):
            _ = dataset[len(dataset) + 100]

    @pytest.mark.parametrize("split", ["train", "val", "test"])
    def test_different_splits(self, split):
        """Test all data splits can be loaded."""
        dataset = SceneVerseDataset({'split': split})
        assert len(dataset) > 0
```

#### 1.2 Model Tests

```python
# tests/test_models/test_llama3d.py
import pytest
import torch
from models.LLM.LLama3d import LLama3D

class TestLLama3D:
    """Test LLama3D model components."""

    @pytest.fixture
    def model(self):
        """Create a test model instance."""
        config = {
            'hidden_size': 256,
            'num_layers': 2,
            'num_heads': 4,
            'vocab_size': 10000
        }
        return LLama3D(config)

    def test_model_initialization(self, model):
        """Test model initializes correctly."""
        assert model is not None
        assert hasattr(model, 'forward')

    def test_forward_pass_shape(self, model):
        """Test forward pass produces correct output shape."""
        batch_size = 2
        seq_len = 10

        input_ids = torch.randint(0, 10000, (batch_size, seq_len))
        point_features = torch.randn(batch_size, 100, 256)

        output = model(input_ids, point_features)

        # Check output shape
        assert output.shape[0] == batch_size
        assert output.shape[1] == seq_len

    def test_gradient_flow(self, model):
        """Test gradients flow through the model."""
        input_ids = torch.randint(0, 10000, (1, 10))
        point_features = torch.randn(1, 100, 256)

        output = model(input_ids, point_features)
        loss = output.sum()
        loss.backward()

        # Check some parameters have gradients
        has_grad = any(p.grad is not None for p in model.parameters())
        assert has_grad, "No gradients computed"

    def test_reproducibility_with_seed(self, model):
        """Test model produces same output with fixed seed."""
        torch.manual_seed(42)
        input_ids = torch.randint(0, 10000, (1, 10))
        point_features = torch.randn(1, 100, 256)

        torch.manual_seed(42)
        output1 = model(input_ids, point_features)

        torch.manual_seed(42)
        output2 = model(input_ids, point_features)

        assert torch.allclose(output1, output2)

    def test_handles_variable_sequence_lengths(self, model):
        """Test model handles different sequence lengths."""
        for seq_len in [5, 10, 20]:
            input_ids = torch.randint(0, 10000, (1, seq_len))
            point_features = torch.randn(1, 100, 256)

            output = model(input_ids, point_features)
            assert output.shape[1] == seq_len
```

#### 1.3 Trainer Tests

```python
# tests/test_trainer/test_trainer.py
import pytest
import torch
from trainer.trainer import ModelingGrounded3DLLM

class TestTrainer:
    """Test trainer functionality."""

    @pytest.fixture
    def trainer(self):
        """Create a minimal trainer for testing."""
        config = {
            'model': {'hidden_size': 128},
            'training': {'batch_size': 2},
            'data': {'max_objects': 50}
        }
        return ModelingGrounded3DLLM(config)

    def test_compute_loss(self, trainer):
        """Test loss computation."""
        predictions = torch.randn(2, 10, 100)
        targets = torch.randint(0, 100, (2, 10))

        loss = trainer.compute_loss(predictions, targets)

        assert isinstance(loss, torch.Tensor)
        assert loss.ndim == 0  # Scalar loss
        assert loss >= 0  # Loss should be non-negative

    def test_evaluation_metrics(self, trainer):
        """Test metric computation."""
        predictions = torch.randn(10, 3)  # 10 boxes
        targets = torch.randn(10, 3)

        metrics = trainer.compute_metrics(predictions, targets)

        assert 'accuracy' in metrics or 'iou' in metrics
        assert all(0 <= v <= 1 for v in metrics.values())

    def test_checkpoint_save_load(self, trainer, tmp_path):
        """Test checkpoint saving and loading."""
        # Save checkpoint
        ckpt_path = tmp_path / "test.ckpt"
        trainer.save_checkpoint(str(ckpt_path))

        assert ckpt_path.exists()

        # Load checkpoint
        new_trainer = ModelingGrounded3DLLM(trainer.config)
        new_trainer.load_checkpoint(str(ckpt_path))

        # Verify model parameters match
        for p1, p2 in zip(trainer.model.parameters(),
                          new_trainer.model.parameters()):
            assert torch.allclose(p1, p2)
```

### Phase 2: Integration Tests (Target: 25% of test suite)

```python
# tests/integration/test_training_pipeline.py
import pytest
import torch
from main_run import train

class TestTrainingPipeline:
    """Integration tests for training pipeline."""

    def test_training_one_epoch(self, tmp_path):
        """Test training runs for one epoch without errors."""
        config = {
            'general': {
                'experiment_name': 'test',
                'save_dir': str(tmp_path)
            },
            'data': {
                'train_split': 'tests/fixtures/mini_train',
                'batch_size': 2
            },
            'training': {
                'max_epochs': 1,
                'val_check_interval': 1.0
            }
        }

        # Should complete without errors
        train(config)

        # Check checkpoint was created
        assert (tmp_path / 'checkpoints').exists()

    def test_evaluation_pipeline(self, tmp_path):
        """Test evaluation runs without errors."""
        # Similar to training test but for evaluation
        pass

    def test_overfitting_single_batch(self):
        """Test model can overfit a single batch (sanity check)."""
        # Train on single batch repeatedly
        # Loss should decrease significantly
        pass
```

### Phase 3: End-to-End Tests (Target: 5% of test suite)

```python
# tests/e2e/test_full_pipeline.py
import pytest

class TestEndToEnd:
    """End-to-end tests for complete workflows."""

    def test_train_and_evaluate(self):
        """Test complete train->evaluate workflow."""
        # 1. Train for a few steps
        # 2. Save checkpoint
        # 3. Load checkpoint
        # 4. Run evaluation
        # 5. Verify results are reasonable
        pass

    def test_inference_on_new_data(self):
        """Test inference on unseen data."""
        # 1. Load pretrained model
        # 2. Run inference on new sample
        # 3. Verify output format
        pass
```

### 1. Testing Framework Configuration

```python
# pytest.ini
[pytest]
testpaths = tests
python_files = test_*.py
python_classes = Test*
python_functions = test_*
addopts =
    -v
    --strict-markers
    --tb=short
    --cov=.
    --cov-report=html
    --cov-report=term-missing
markers =
    slow: marks tests as slow (deselect with '-m "not slow"')
    gpu: marks tests that require GPU
    integration: marks integration tests

# Ignore warnings from dependencies
filterwarnings =
    ignore::DeprecationWarning
    ignore::PendingDeprecationWarning
```

### 2. Test Fixtures

```python
# tests/conftest.py
"""Shared pytest fixtures."""
import pytest
import torch

@pytest.fixture(scope="session")
def device():
    """Get available device (GPU if available, else CPU)."""
    return torch.device('cuda' if torch.cuda.is_available() else 'cpu')

@pytest.fixture(autouse=True)
def reset_seed():
    """Reset random seed before each test."""
    torch.manual_seed(42)
    import numpy as np
    np.random.seed(42)
    import random
    random.seed(42)

@pytest.fixture(scope="session")
def test_data_dir(tmp_path_factory):
    """Create temporary directory for test data."""
    return tmp_path_factory.mktemp("test_data")
```

### 3. Mock Data Generation

```python
# tests/fixtures/generate_test_data.py
"""Generate minimal test datasets."""
import torch
import json

def generate_mini_dataset(output_dir, num_samples=10):
    """Generate a minimal dataset for testing."""
    output_dir.mkdir(parents=True, exist_ok=True)

    data = []
    for i in range(num_samples):
        sample = {
            'scene_id': f'scene_{i:03d}',
            'point_cloud': torch.randn(1000, 6).tolist(),  # XYZ + RGB
            'language': f'Find the {i}th object',
            'target_id': i
        }
        data.append(sample)

    with open(output_dir / 'test_data.json', 'w') as f:
        json.dump(data, f)
```

### Week 1: Setup and Dataset Tests
- [x] Set up pytest configuration
- [ ] Create test directory structure
- [ ] Generate test fixtures
- [ ] Write dataset unit tests
- [ ] Achieve 30% coverage on datasets/

### Week 2: Model Tests
- [ ] Write model unit tests
- [ ] Test model components
- [ ] Test forward/backward passes
- [ ] Achieve 40% coverage on models/

### Week 3: Trainer Tests
- [ ] Write trainer unit tests
- [ ] Test loss computation
- [ ] Test metric calculation
- [ ] Achieve 30% coverage on trainer/

### Week 4: Integration and CI
- [x] Write integration tests
- [ ] Set up GitHub Actions CI
- [ ] Add coverage reporting
- [ ] Document testing procedures

## Continuous Integration Setup

```yaml
# .github/workflows/test.yml
name: Tests

on: [push, pull_request]

jobs:
  test:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        python-version: ['3.10']

    steps:
    - uses: actions/checkout@v3

    - name: Set up Python
      uses: actions/setup-python@v4
      with:
        python-version: ${{ matrix.python-version }}

    - name: Install dependencies
      run: |
        pip install -r requirements.txt
        pip install pytest pytest-cov

    - name: Run tests
      run: |
        pytest tests/ -v --cov=. --cov-report=xml

    - name: Upload coverage
      uses: codecov/codecov-action@v3
      with:
        file: ./coverage.xml
```

### 1. Test Naming Convention
```python
def test_<component>_<expected_behavior>_<condition>():
    """Test that component does expected_behavior when condition."""
    pass
```

### 2. Arrange-Act-Assert Pattern
```python
def test_example():
    # Arrange: Set up test data
    model = create_model()
    input_data = create_input()

    # Act: Execute the code being tested
    output = model(input_data)

    # Assert: Verify the results
    assert output.shape == expected_shape
```

### 3. Test Independence
- Each test should be independent
- Use fixtures for shared setup
- Clean up resources after tests

### 4. Meaningful Assertions
```python
# ❌ Bad
assert output is not None

# ✅ Good
assert output.shape == (batch_size, seq_len, hidden_dim)
assert torch.all(output >= 0), "Output should be non-negative"
```

## Expected Benefits

✅ **Regression Prevention**: Catch bugs before they reach production
✅ **Refactoring Confidence**: Safely modify code with test coverage
✅ **Documentation**: Tests serve as usage examples
✅ **Faster Debugging**: Identify issues quickly with failing tests
✅ **Reproducibility**: Verify consistent behavior
✅ **Collaboration**: Contributors can verify their changes don't break existing functionality

## Success Metrics

- **Code Coverage**: Achieve 60%+ overall coverage
- **Test Count**: 100+ unit tests, 20+ integration tests
- **CI/CD**: All tests pass on every commit
- **Performance**: Test suite runs in < 5 minutes

**Comment by zc2023** [comment]
Thank you for your suggestion. We have verified the usability within our internal code repository.
